homeassistant/components/letpot/__init__.py

```diff
@@ -0,0 +1,95 @@
+"""The LetPot integration."""
+
+from __future__ import annotations
+
+import asyncio
+
+from letpot.client import LetPotClient
+from letpot.converters import CONVERTERS
+from letpot.exceptions import LetPotAuthenticationException, LetPotException
+from letpot.models import AuthenticationInfo
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_EMAIL, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import (
+    CONF_ACCESS_TOKEN_EXPIRES,
+    CONF_REFRESH_TOKEN,
+    CONF_REFRESH_TOKEN_EXPIRES,
+    CONF_USER_ID,
+)
+from .coordinator import LetPotDeviceCoordinator
+
+PLATFORMS: list[Platform] = [Platform.TIME]
+
+type LetPotConfigEntry = ConfigEntry[list[LetPotDeviceCoordinator]]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: LetPotConfigEntry) -> bool:
+    """Set up LetPot from a config entry."""
+
+    auth = AuthenticationInfo(
+        access_token=entry.data[CONF_ACCESS_TOKEN],
+        access_token_expires=entry.data[CONF_ACCESS_TOKEN_EXPIRES],
+        refresh_token=entry.data[CONF_REFRESH_TOKEN],
+        refresh_token_expires=entry.data[CONF_REFRESH_TOKEN_EXPIRES],
+        user_id=entry.data[CONF_USER_ID],
+        email=entry.data[CONF_EMAIL],
+    )
+    websession = async_get_clientsession(hass)
+    client = LetPotClient(websession, auth)
+
+    if not auth.is_valid:
+        try:
+            auth = await client.refresh_token()
+            hass.config_entries.async_update_entry(
+                entry,
+                data={
+                    CONF_ACCESS_TOKEN: auth.access_token,
+                    CONF_ACCESS_TOKEN_EXPIRES: auth.access_token_expires,
+                    CONF_REFRESH_TOKEN: auth.refresh_token,
+                    CONF_REFRESH_TOKEN_EXPIRES: auth.refresh_token_expires,
+                    CONF_USER_ID: auth.user_id,
+                    CONF_EMAIL: auth.email,
+                },
+            )
+        except LetPotAuthenticationException as exc:
+            raise ConfigEntryAuthFailed from exc
+
+    try:
+        devices = await client.get_devices()
+    except LetPotAuthenticationException as exc:
+        raise ConfigEntryAuthFailed from exc
```

Comment:
You can only raise ConfigEntryAuthFailed if you have a reauth flow, so let's make these ConfigEntryErrors for now

Reply:
For places where I've also added `except LetPotException`: should I just remove the authentication specific lines for now then? (`LetPotException` is the base class)

Reply:
With Not ready it will retry, and that's not correct I think